### ISSUE TITLE
CMA_diagonal /VdK only works with adapt_sigma = FALSE?

### ISSUE THREAD

**Issue by adder** [CMA-ES/pycma#56]
Hey,

I was experimenting with CMA-ES with high dimensional parameter spaces.
For speed, I tried to use the CMA_diagonal = TRUE option.
I noticed it only works when I also set adapt_sigma is to FALSE.
Is this normal behaviour?
If so, is there a reason when do not allow sigma to change with a diagonal covariance matrix?
I also noticed that the default setting for adapt_sigma in the VdK sampler is also zero?
What is the advised procedure when using this sampler?
Setting sigma fixed for the entire run, manually restarting the run with a stepwise lower sigma or something else?

Thanks for the nice piece of software and algorithm!
Best Regards

**Comment by nikohansen** [comment]
[Here](https://github.com/CMA-ES/pycma/blob/master/cma/test.py#L285) you can see a usecase for using the VD and VkD sampler.

Setting `CMA_diagonal` to `True` should work fine. You might have discovered a bug, so it would be interesting if you could give more information on the specific calling sequence and what "only works" mean.

**Comment by adder** [comment]
Hey
Regarding `CMA_diagonal`.
I did a little more digging. I have the error with the different provided interfaces. eg:
```
res = cma.fmin(func,x, 0.5, {'maxiter': 10, 'CMA_diagonal' : True})
```

However I noticed, I only have the error if the parameter vector is rather large (10800) in another setting with less parameters (180) everything runs fine.
When setting `AdaptSigma = False` , I never get the error.  

The error I get with a large parameter space is and `AdaptSigma = True` is :
```
(15_w,31)-aCMA-ES (mu_w=8.9,w_1=20%) in dimension 10800 (seed=1016509, Fri Mar  2 07:38:06 2018)
   Covariance matrix is diagonal
Error in py_call_impl(callable, dots$args, dots$keywords) : 
  TypeError: '<' not supported between instances of 'list' and 'float'

Detailed traceback: 
  File "/usr/lib/python3.6/site-packages/cma/evolution_strategy.py", line 4045, in fmin
    es.tell(X, fit)  # prepare for next iteration
  File "/usr/lib/python3.6/site-packages/cma/evolution_strategy.py", line 2638, in tell
    self.adapt_sigma.update(self, function_values=function_values)
  File "/usr/lib/python3.6/site-packages/cma/sigma_adaptation.py", line 376, in update
    z = sum(f_vals < f_vals[1]) - sum(f_vals < f_vals[0])
```
Regarding the VkD sampler. I did base myself on the test file and got it succesfully to run.
However, I didn't manage to set `k` manually.
I tried the folowing
```
opts = {'CMA_sampler_options' : {'kadapt' : False, 'k_init' : 1}}
rgs =cma.restricted_gaussian_sampler.GaussVkDSampler.extend_cma_options(opts)
res = cma.fmin(func,x, 0.5, rgs)                                                        
```
But I got the same results as without manually setting `k`, so I guess maybe the options didn't pass properly?       

I also noticed that `AdaptSigma` is by default set to `False` in VkD an I was wondering why?
Maybe I'm not fully grasping what this option implies?

Best regards

**Comment by nikohansen** [comment]
> The error I get with a large parameter space is and AdaptSigma = True is

This looks like a bug. I cannot replicate the error, but I have pushed what I believe is a fix to the `master` branch.

> But I got the same results as without manually setting k, so I guess maybe the options didn't pass properly?

AFAICS the parameter is passed and effective. Whether we expect to see an effect certainly depends on the function to be optimized.

> I also noticed that AdaptSigma is by default set to False in VkD an I was wondering why?

It must be set to `False`, because the sampler takes care of sigma adaptation internally.

**Comment by nikohansen** [comment]
FTR, default setting
```python
opts = {'maxiter': 2}
rgs = cma.restricted_gaussian_sampler.GaussVkDSampler.extend_cma_options(opts)
es = cma.fmin2(cma.ff.sphere, x, 0.5, rgs)[1]
cma.s.pprint(es.sm.__dict__)
```
```
(15_w,31)-CMA-ES (mu_w=8.9,w_1=20%) in dimension 10800 (seed=1117021, Fri Mar  2 11:40:32 2018)
Iterat #Fevals   function value  axis ratio  sigma  min&max std  t[m:s]
    1     31 1.325094976048201e+04 1.0e+00 5.00e-01  5e-01  5e-01 0:00.3
    2     62 1.230540811354117e+04 1.0e+00 5.00e-01  5e-01  5e-01 0:00.7
termination on maxiter=2 (Fri Mar  2 11:40:33 2018)
final/bestever f-value = 1.089285e+04 1.089285e+04
incumbent solution: [ 0.88556779  0.95730695  0.93826249  1.12019828  0.86282234  0.97264591
  0.92420615  0.95224572 ...]
std deviations: [ 0.48169219  0.48153981  0.48148353  0.48158027  0.4814567   0.48128749
  0.48156293  0.48180887 ...]
{'D': array([ 1.00019983,  0.99988342,  0.99976655, ...,  1.00026271,
        0.99989231,  1.00034351]),
 'N': 10800,
 'S': array([ 0.,  0.,  0., ...,  0.,  0.,  0.]),
 'U': ...,
 'V': array([], shape=(0, 10800), dtype=float64),
 '_debug': False,
 '_mueff': 8.900967696718343,
 '_parameters': {'c1': 9.253732310110796e-05,
                 'cc': 0.009619632170780126,
                 'cmu': 0.001296732780063776},
 'accepted_slowdown': 3.0,
 'cs': 0.3,
 'ds': 3.999722222222222,
 'dx': array([-0.08584175, -0.33752614,  0.05264434, ..., -0.61236737,
        0.00106194,  0.33927548]),
 'ema_log_d': <cma.restricted_gaussian_sampler.ExponentialMovingAverage object at 0x11f12cdd8>,
 'ema_log_s': <cma.restricted_gaussian_sampler.ExponentialMovingAverage object at 0x11f12cba8>,
 'ema_log_sigma': <cma.restricted_gaussian_sampler.ExponentialMovingAverage object at 0x11f12c320>,
 'factor_diag_slope': 2,
 'factor_sigma_slope': 0.1,
 'flg_injection': True,
 'itr_after_k_inc': 2,
 'k': 0,
 'k_active': 0,
 'k_adapt_decay': 9.259259259259259e-05,
 'k_adapt_factor': 1.414,
 'k_adapt_wait': 21599.0,
 'k_dec_cond': 30.0,
 'k_inc_cond': 30.0,
 'kadapt': True,
 'kmax': 10799,
 'kmin': 0,
 'last_log_cond_corr': array([ 0.,  0.,  0., ...,  0.,  0.,  0.]),
 'last_log_d': array([ 0.00039962, -0.00023317, -0.00046696, ...,  0.00052535,
       -0.00021538,  0.00068691]),
 'last_log_sigma': -0.03750260434752412,
 'lndiag_change': array([-0.04908504,  0.00507484, -0.39735125, ...,  0.13905797,
       -0.04606272,  0.25670529]),
 'lnlambda_change': array([ 0.,  0.,  0., ...,  0.,  0.,  0.]),
 'lnsigma_change': array([-78.3925407]),
 'opt_conv': 0.0014351851851851852,
 'pc': array([-0.09392051, -0.03624764, -0.05028017, ..., -0.232655  ,
       -0.0870427 , -0.04146652]),
 'ps': -0.15,
 'randn': <built-in method randn of mtrand.RandomState object at 0x10d83f5a0>,
 'sigma': 0.9631919092310912,
 'sigma_fac': 1.0}
```
versus setting ``'kadapt' : False``
```python
opts = {'maxiter': 2, 'CMA_sampler_options' : {'kadapt' : False, 'k_init' : 1}}
rgs = cma.restricted_gaussian_sampler.GaussVkDSampler.extend_cma_options(opts)
es = cma.fmin2(cma.ff.sphere, x, 0.5, rgs)[1]
cma.s.pprint(es.sm.__dict__)
```
```
(15_w,31)-CMA-ES (mu_w=8.9,w_1=20%) in dimension 10800 (seed=1166664, Fri Mar  2 11:42:11 2018)
Iterat #Fevals   function value  axis ratio  sigma  min&max std  t[m:s]
    1     31 1.329669722700274e+04 1.0e+00 5.00e-01  5e-01  5e-01 0:00.4
    2     62 1.277728237917903e+04 1.0e+00 5.00e-01  5e-01  5e-01 0:00.7
termination on maxiter=2 (Fri Mar  2 11:42:12 2018)
final/bestever f-value = 1.094257e+04 1.094257e+04
incumbent solution: [ 1.04940899  0.78097936  1.10380503  0.96597019  0.72508475  0.99470855
  1.02232392  1.20151038 ...]
std deviations: [ 0.48164431  0.48167119  0.48149654  0.4815904   0.48165641  0.481497
  0.48184009  0.48159053 ...]
{'D': array([ 0.99996006,  1.00000225,  0.99977246, ...,  0.99996342,
        0.9998674 ,  0.99976971]),
 'N': 10800,
 'S': array([ 2.60278697,  0.        ,  0.        , ...,  0.        ,
        0.        ,  0.        ]),
 'U': ...,
 'V': array([[ 0.01038557,  0.01087735,  0.00402801, ..., -0.01362898,
         0.01784784,  0.00482696]]),
 '_debug': False,
 '_mueff': 8.900967696718343,
 '_parameters': {'c1': 6.170001882754827e-05,
                 'cc': 0.007854935953115611,
                 'cmu': 0.0006487532635421981},
 'cs': 0.3,
 'ds': 3.999722222222222,
 'dx': array([ 0.35603131, -0.12251304,  0.46783113, ..., -0.39445433,
        0.28416715,  0.42968863]),
 'flg_injection': True,
 'k': 1,
 'k_active': 1,
 'kadapt': False,
 'pc': array([ 0.03763009, -0.1625703 ,  0.07824293, ...,  0.00489175,
       -0.09550043,  0.11960844]),
 'ps': -0.15,
 'randn': <built-in method randn of mtrand.RandomState object at 0x10d83f5a0>,
 'sigma': 0.9631919092310912,
 'sigma_fac': 1.0}
```

**Comment by adder** [comment]
Hey thanks for the quick fix!

I pulled the latest code from github and `CMA_diagonal` seems to work now!
I did not looked further into the `VdK` sampler with `kadapt = False` and `k_init =0`.
I actually wanted to use it to replicate the behaviour of `CMA_diagonal` to do CMA-ES with only a diagonal covariance matrix to speed up the optimization but since the `CMA_diagonal` setting is working now, I can use this instead.
On the other hand, the default `VdK` sampler with adaptive k seems also pretty fast out of the box. 
But I guess it depends on the optimization problem at hand like you said.
I'm definitly also going to exploring this setting more.

Thanks again for the great and fast help!

Best regards
